src/extractor/presets/react.ts

```diff
@@ -1,20 +1,85 @@
-import type { ExtractorRegExp } from '../regexp';
-
-module.exports = <ExtractorRegExp>{
-  references: {
-    hook: /\s([a-zA-Z0-9_]+)\s?=\s??useTranslate\(\)/g,
-  },
-  extraction: [
-    // React component (JSX)
-    '<\\s*T [^>]*?keyName={?["\'`]%string%["\'`]}?',
-    '<\\s*T[^>]*?>%string%</T>}?',
-
-    // React component (Plain React.createElement)
-    // fixme: nested objects would cancel the `[^}]*?` segment
-    'React\\.createElement\\(\\s*T,\\s*{[^}]*?keyName:\\s*["\'`]%string%["\'`]',
-
-    // Hook usage
-    '[^A-Za-z0-9_]$hook$\\(["\'`]%string%["\'`],?',
-    '[^A-Za-z0-9_]$hook$\\({(?:\\n|.*?)*?key:\\s*[\'`"]%string%[\'`"]',
-  ],
+import {
+  isConsumable,
+  greedilyConsumeBlock,
+  extractString,
+  extractObject,
+} from '../languages/javascript';
+
+type CreateElementBlock = {
+  component: string;
+  props: Record<string, string>;
+  children: string | null;
 };
+
```

Comment:
Are you really sure about this? The previous solution looked much more simple and mantainable. I would still suggest to keep the regexes with parallelization / rewrite it to state machines with parallelization. This solution looks like some kind of hybrid and I am conserned, it's not more maintainable then the regexes.

Reply:
It is more complex than regex, but it's close to impossible to achieve levels of accuracy I'm happy with, without horrendous looking regexes and a lot of nested conditions. Regexes suck at understanding context, yet context is an absolute must for having any form of meaningful extraction.

For example: to extract this object: `{ key1: "value}", key2: { key3 } }`, there are 2 "traps" for the regex: the `}` embedded in the first value, and the nested object which *is* part of the object. Greedy matcher may go out of the object boundary, non-greedy marker will stop prematurely. In both cases we failed to properly extract the object. This is why in `javascript.ts`, the greedy consume helper keeps track of "depth": it allows to consume nested structures while keeping an accurate indicator of where we're at in the consumption.

This is why I'm moving away from regex here: for built-in extractors I expect relatively high accuracy at a reasonable efficiency: and regexes provide neither of those. They are slow by nature, complex ones are prone to a whole class of program errors and runaway forcing the CLI to abort, for quite deceiving accuracy in these complex tasks code extraction. Regexes are capable of code extraction, if one is willing to spend extra time having the infrastructure capable of doing so, and I don't want to reinvent TextMate grammar.

Concerning maintainability concerns in the React extractor: it looks heavy because I still have some bits of code analysis tucked in there (argument decode logic). Once all have been extracted, it'll only be about cherry-picking bits of code, calling some procedures from `javascript.ts`, and boom done. The JavaScript part is quite huge, with somewhat long procedures, but it can be easily tested and makes maintenance in fact much easier imho.

The fact the analysis logic is split from the extraction routine means we can test the analysis first, which is a bunch of procedures working on specific tasks that can be tested on a case-per-case, and then if this works we know we already have a solid foundation for the React extractor, but also for all the other JS-based frameworks!

----

I am not entirely against bringing back a regex helper, but it'll need a lot more consideration compared to what it was before. I'm concerned the amount of concerns and steps to describe might make it a more sensible choice to let people who don't need more than regex "roll their own". In all cases, I feel like this is a task way too large to fit in this already large PR.